core/corehttp/ipns_hostname.go

```diff
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
 			defer cancel()
 
 			host := strings.SplitN(r.Host, ":", 2)[0]
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+			if p, err := n.Namesys.Resolve(ctx, host, 0); err == nil {
+				r.URL.Path = p.String() + r.URL.Path
```

Comment:
On Sat, May 16, 2015 at 02:57:34PM -0700, Knut Ahlers wrote:
> … which is redirected by the mux (not by any IPFS code)…

Ah, I see.  So having the trailing double-slash is a problem for
anything that's getting DNS-link resolution.

> Regarding domains:
> - My public gateway (ipfs.hub.luzifer.io) has no `dnslink` TXT entry
>   so this does not apply to those gateways.
> - swagger.luzifer.io is a website. Not a gateway. If someone tries
>   to use it as a gateway it should fail by design. Accessing content
>   outside `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA`
>   would be gateway functionality…

So that sounds like “the way you distinguish between absolute and
relative paths is whether or not the host has a DNS-link TXT entry”.
I'm fine with that.  And then “do whatever we need to do to avoid
upsetting the muxer,” and the TrimRight makes sense there.

I'd rather not take the /ipns/ approach [1], because then we're
hitting the DNS entry twice (one for “is there a DNS-link” and again
for “resolve this DNS-link-based /ipns/… path”).  And because we've
rolled back my proposed distinction between DNS and DHT links [2],
there's some extra multihash checking that we wouldn't need.  Neither
of those are big deals, but I don't like pulling in logic we know we
won't need here.

[1]: https://github.com/ipfs/go-ipfs/pull/1208/files#r30463056
[2]: https://github.com/ipfs/go-ipfs/pull/1208#issuecomment-100754017